presto-main/src/main/java/com/facebook/presto/cost/CachingCostProvider.java

```diff
@@ -40,16 +39,16 @@
     private final Optional<Memo> memo;
     private final Lookup lookup;
     private final Session session;
-    private final Supplier<Map<Symbol, Type>> types;
+    private final Supplier<TypeProvider> types;
```

Comment:
From API perspective, there isn't much difference between `Supplier<TypeProvider>` and `TypeProvider` itself. `Supplier<Map>` was used so that we have up-to-date information when `CostProvider` is called from a Rule, which just allocated some new symbols before asking for cost.
Now that you introduced a new `TypeProvider` type, this could be handled with a `TypeProvider` that isn't a wrapper around static (immutable) collection, but e.g. delegates to `SymbolAllocator` instead.